### PR TITLE
test: Accuracy test improvement (Part 3.1): Extend accuracy test suite with LLM API and initial implementation of `trtllm-eval`

### PR DESCRIPTION
**Pull request by syuoni**
This PR:
* Extends accuracy test suite with LLM API;
* Removes `exampes/test_pytorch.py` and ports all test cases to accuracy test suite;
* Introduces `trtllm-eval`.

-----
Example usage of `trtllm-eval`:

```bash
$ trtllm-eval --model /home/scratch.trt_llm_data/llm-models/llama-3.1-model/Meta-Llama-3.1-8B cnn_dailymail --num_samples 512
......
[03/31/2025-09:54:57] [TRT-LLM] [I] TRTLLM execution time: 17.161 seconds.
[03/31/2025-09:54:58] [TRT-LLM] [I]   rouge1: 24.667
[03/31/2025-09:54:58] [TRT-LLM] [I]   rouge2: 8.048
[03/31/2025-09:54:58] [TRT-LLM] [I]   rougeL: 16.771
[03/31/2025-09:54:58] [TRT-LLM] [I]   rougeLsum: 20.997
```
```bash
$ trtllm-eval --model /home/scratch.trt_llm_data/llm-models/llama-3.1-model/Meta-Llama-3.1-8B mmlu --dataset_path /home/scratch.trt_llm_data/llm-models/datasets/mmlu --num_samples 512
......
[03/31/2025-09:54:07] [TRT-LLM] [I] Average accuracy 100.00 (9) - geography
[03/31/2025-09:54:07] [TRT-LLM] [I] Average accuracy 77.78 (36) - politics
[03/31/2025-09:54:07] [TRT-LLM] [I] Average accuracy 72.22 (18) - psychology
[03/31/2025-09:54:07] [TRT-LLM] [I] Average accuracy 88.89 (18) - culture
[03/31/2025-09:54:07] [TRT-LLM] [I] Average accuracy 70.37 (27) - law
[03/31/2025-09:54:07] [TRT-LLM] [I] Average accuracy 60.49 (162) - STEM
[03/31/2025-09:54:07] [TRT-LLM] [I] Average accuracy 64.96 (117) - humanities
[03/31/2025-09:54:07] [TRT-LLM] [I] Average accuracy 78.70 (108) - social sciences
[03/31/2025-09:54:07] [TRT-LLM] [I] Average accuracy 64.29 (126) - other (business, health, misc.)
[03/31/2025-09:54:07] [TRT-LLM] [I] MMLU weighted average accuracy: 66.28 (513)
```

### CODE REVIEW

File: tensorrt_llm/llmapi/llm_utils.py [hunk 1]
```diff
@@ -343,6 +343,84 @@ def _download_hf_model(self):
 
             assert self.speculative_model_obj.is_local_model
 
+    def _update_from_hf_quant_config(self) -> bool:
```
Comment (syuoni) [inline tensorrt_llm/llmapi/llm_utils.py:346]: Hi @kaiyux ,

I separate these logics to a dedicated method, also slightly generalize the logics. Could you please review this part?

File: tensorrt_llm/llmapi/llm_utils.py [hunk 2]
```diff
@@ -615,6 +654,10 @@ def __call__(self) -> Tuple[Path, Union[Path, None]]:
                 logger.warning(
                     "QuantConfig for pytorch backend is ignored. You can load"
                     "quantized model with hf_quant_config.json directly.")
+            # Currently, this is to make updated quant_config visible by llm.args.quant_config
+            # TODO: Unify the logics with those in tensorrt_llm/_torch/model_config.py
+            self.model_loader._update_from_hf_quant_config()
```
Comment (syuoni) [inline tensorrt_llm/llmapi/llm_utils.py:665]: Hi @QiJune ,

There are some duplication between here and `tensorrt_llm/_torch/model_config.py` that reads quant_config from pre-quantized checkpoint. I call `_update_from_hf_quant_config` here so that users can access updated quant_config via `llm.args.quant_config`.

I'm wondering if we can unify these two places, possibly by passing the updated quant_config to PyTorch engine. Do you have any concerns?

File: tests/integration/defs/accuracy/test_cli.py [hunk 3]
```diff
@@ -14,16 +14,18 @@
 # limitations under the License.
 import pytest
```
Comment (syuoni) [inline tests/integration/defs/accuracy/test_cli.py]: Hi @byshiue ,

I renamed `test_accuracy.py` to `test_cli.py`, so that the tests can be clearly distinguished from those in `test_llm_api.py` and `test_llm_api_pytorch.py`.

Unfortunately,  this causes massive renamed test cases. The actual new test cases are only about 10, which are all started with `test_llm_api.py` or `test_llm_api_pytorch.py`.

Comment (byshiue) [reply tests/integration/defs/accuracy/test_cli.py]: LGTM

File: tensorrt_llm/quantization/functional.py [hunk 4]
```diff
@@ -1190,10 +1190,11 @@ def postprocess_fp8_rowwise(tllm_key, weights, **kwargs):
             tllm_key.replace("weight", "per_channel_scale"): scales
         }
     else:
+        x = torch.cat(weights, dim=0).to(torch.float32)
         clamp_val = config.quantization.clamp_val
-        weights = torch.cat(weights, dim=0)
-        # activation range bound.
-        x = weights.to(torch.float32).clamp(clamp_val[0], clamp_val[1])
+        if clamp_val is not None:
```
Comment (byshiue) [inline tensorrt_llm/quantization/functional.py:1195]: When `clamp_val` would be `None`?

**Review by byshiue — APPROVED** [review]
LGTM

File: tensorrt_llm/evaluate/cnn_dailymail.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+import evaluate
+
+from .._torch import LLM as PyTorchLLM
+from ..llmapi import LLM, RequestOutput
+from ..logger import logger
+from ..sampling_params import SamplingParams
+from .interface import Evaluator
+
+
+class CnnDailymail(Evaluator):
```
Comment (QiJune) [inline tensorrt_llm/evaluate/cnn_dailymail.py:28]: Cnn daily is a dataset name. How about `Summarization`?

File: tensorrt_llm/evaluate/mmlu.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+import pandas as pd
+
+from .._torch import LLM as PyTorchLLM
+from ..llmapi import LLM, RequestOutput
+from ..logger import logger
+from ..sampling_params import SamplingParams
+from .interface import Evaluator
+
+
+class Mmlu(Evaluator):
```
Comment (QiJune) [inline tensorrt_llm/evaluate/mmlu.py]: Mmlu -> MMLU

File: tensorrt_llm/llmapi/llm_utils.py (hunk 2 above)
Comment (QiJune) [reply tensorrt_llm/llmapi/llm_utils.py:665]: I know trtllm-bench tool will read quant_config directly: https://github.com/NVIDIA/TensorRT-LLM/blob/727d78e785da96ce8ac28f920841819efaeee220/tensorrt_llm/bench/benchmark/utils/general.py#L109-L112

It's not through `LLM` API.

File: tests/integration/defs/accuracy/test_cli.py [hunk 7]
```diff
@@ -14,16 +14,18 @@
...
 
+from tensorrt_llm.llmapi import (EagleDecodingConfig, LookaheadDecodingConfig,
+                                 MedusaDecodingConfig)
 from tensorrt_llm.quantization import QuantAlgo
 
 from ..conftest import (llm_models_root, skip_no_nvls, skip_pre_ada,
                         skip_pre_blackwell, skip_pre_hopper)
-from .accuracy_core import (AccuracyTestHarness, CnnDailymail, Humaneval, Mmlu,
-                            PassKeyRetrieval64k, PassKeyRetrieval128k,
+from .accuracy_core import (CliAccuracyTestHarness, CnnDailymail, Humaneval,
```
Comment (QiJune) [inline tests/integration/defs/accuracy/test_cli.py]: It takes me some time to understand `test_cli.py` means that convert -> build -> run flow. But I don't find a better name yet.

File: tensorrt_llm/llmapi/llm_utils.py (hunk 2 above)
Comment (syuoni) [reply tensorrt_llm/llmapi/llm_utils.py:665]: Yes. Once we unify them, `trtllm-bench` can also rely on `llm.args.quant_config`, which avoids repeated loading `ModelConfig`.

File: tensorrt_llm/llmapi/llm_utils.py [hunk 8]
```diff
@@ -353,46 +431,7 @@ def _load_model_from_hf(self):
...
-                            f"will overwrite {hf_kv_cache_quant_algo} from {hf_quant_config_path}."
-                        )
-                    else:
-                        raise ValueError(
-                            f"User specified kv_cache_quant_algo {user_quant_config.kv_cache_quant_algo}, "
-                            f"while it's {hf_kv_cache_quant_algo} in {hf_quant_config_path}."
-                        )
-        else:
-            already_quantized = False
+        pre_quantized = self._update_from_hf_quant_config()
```
Comment (kaiyux) [inline tensorrt_llm/llmapi/llm_utils.py]: nit: `pre_quantized` -> `prequantized`

File: tensorrt_llm/llmapi/llm_utils.py [hunk 9]
```diff
@@ -343,6 +343,84 @@ def _download_hf_model(self):
...
+                    )
+
+            for key, value in hf_quant_config.items():
+                logger.info(f"Setting {key}={value} from HF quant config.")
+                setattr(quant_config, key, value)
+
+            return True
+
+        hf_config_path = f"{self._model_dir}/config.json"
+        if os.path.exists(hf_config_path):
```
Comment (kaiyux) [inline tensorrt_llm/llmapi/llm_utils.py:407]: I guess we don't need to check when `hf_quant_config.json` and `config.json` both exist and they have conflict information?

File: tensorrt_llm/llmapi/llm_utils.py [hunk 10]
```diff
@@ -343,6 +343,84 @@ def _download_hf_model(self):
...
+
+            if hf_quant_config is not None:
+                # DeepSeek V3 FP8 ckpt
+                if hf_quant_config.get(
+                        "quant_method") == "fp8" and hf_quant_config.get(
+                            "weight_block_size", []):
+                    quant_config.quant_algo = QuantAlgo.FP8_BLOCK_SCALES
+                    quant_config.exclude_modules = ["*eh_proj"]
+
+                return True
```
Comment (kaiyux) [inline tensorrt_llm/llmapi/llm_utils.py:426]: IIUC this means when `quant_method` is not `fp8`, we are not going to set `quant_algo`, but still return `True`. Is that going to lead to some issues? Should we add a warning or maybe even NotImplementedError?

File: tensorrt_llm/llmapi/llm_utils.py (hunk 1 above)
Comment (kaiyux) [reply tensorrt_llm/llmapi/llm_utils.py:346]: Added some comments, thanks!

File: tensorrt_llm/quantization/functional.py (hunk 4 above)
Comment (syuoni) [reply tensorrt_llm/quantization/functional.py:1195]: According to the plugin implementation, `clamp_val` is optional, see [here](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/quantization/functional.py#L460) and [here](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/quantization/functional.py#L504).

I've verified everything works well when setting `clamp_val=None`.

File: tensorrt_llm/evaluate/cnn_dailymail.py (hunk 5 above)
Comment (syuoni) [reply tensorrt_llm/evaluate/cnn_dailymail.py:28]: There could be multiple summarization datasets. In TRT-LLM, we already have another one [squality from tau/zero_scrolls](https://github.com/NVIDIA/TensorRT-LLM/blob/main/examples/summarize.py#L88). So, `Summarization` might be a bit vague.

Same logic applies to MMLU, I suppose we should not rename it as `QA` or `MultipleChoice`, right?

File: tensorrt_llm/llmapi/llm_utils.py (hunk 9 above)
Comment (syuoni) [reply tensorrt_llm/llmapi/llm_utils.py:407]: Yes. Normally it should be the checkpoint provider's responsibility to ensure the configs are correct and consistent.

File: tensorrt_llm/evaluate/mmlu.py (hunk 6 above)
Comment (syuoni) [reply tensorrt_llm/evaluate/mmlu.py]: Done. Thanks!

File: tensorrt_llm/llmapi/llm_utils.py (hunk 8 above)
Comment (syuoni) [reply tensorrt_llm/llmapi/llm_utils.py]: Renamed, thanks!

File: tensorrt_llm/llmapi/llm_utils.py (hunk 10 above)
Comment (syuoni) [reply tensorrt_llm/llmapi/llm_utils.py:426]: Good suggestion! Added `NotImplementedError`.

File: tests/integration/defs/accuracy/test_cli.py (hunk 7 above)
Comment (syuoni) [reply tests/integration/defs/accuracy/test_cli.py]: As offline discussed, renamed to `test_cli_flow.py`. Thanks!

**Comment by syuoni** [comment]
Hi all, 

The remaining threads don't look too risky to me. I'm resolving all threads to merge this PR. Please let me know if you still have concerns. 

Thanks!
